librad/src/meta/entity/data.rs

```diff
@@ -55,7 +57,50 @@ where
     ordered.serialize(serializer)
 }
 
-/// Raw data for an `Entity
+/// The different kinds of `Entity` types
+#[derive(Clone, Copy, PartialEq, Eq, Deserialize, Serialize, Debug)]
+pub enum EntityKind {
+    User,
+    Project,
+}
+
+/// Information payload for all possible `Entity` types
+///
+/// This is used so we can deserialize `EntityData` without knowing the actual
+/// entity type: deserializing as `EntityData<EntityInfo>` the `info` property
+/// will contain this payload
+///
+/// Care has been taken to ensure that the same data can be deserialized both
+/// "generically" (as `EntityData<EntityInfo>`) and using its specific type
+#[derive(Clone, Deserialize, Serialize, Debug)]
+#[serde(untagged)]
+pub enum EntityInfo {
+    User(UserInfo),
```

Comment:
I think I see where my disconnection is coming from. I was expecting something more like:
```rust
pub enum EntityType<ST> {
  User(Entity<UserInfo, ST>),
  Project(Entity<ProjectInfo, ST>),
}
```

I think this would avoid having `EntityKind` and the extra fields on `User` and `Project`.

Reply:
To be fair, `EntityKind` is probably not needed anymore.
Initially I was using it to implement the "manual internally tagged: trick, then I noticed it could not work and switched to the current approach with no in-memory "tag" at all (the tag is only in the blob).

If we find it confusing I can try to remove it.
It could be _convenient_ to have some kind of tag at run time that tells us "what this entity is" transparently, both for the enum and the specific types, but this is not the foundation of how types are handled.

About the general design, putting the union type at the "info" level has the advantage of naturally accessing an Entity as an Entity, regardless of the type.
I mean, also the signature and verification steps work on `GenericEntity`, with no additional "unwrap the enum" code needed.

I find this a good thing.
Instinctively I would have also started with the "entity level" enum, but I choose this other approach on purpose because:

* it solves the "serialization compatibility problem"
* it allows us to transparently access all the entity-level methods also on the union type
* it allowed a nice simplification of the "check the specific invariants at build time" part, eliminating the `EntityBuilder` trait
* this last thing is also related to the removal of the requirement to implement `Default` for `info` which was annoying

Reply:
I see -- this is making more sense now. I think `EntityKind` should be needed because we'll know the "kind" when we inspect what `EntityInfo` it has.

So I think the `EntityInfo` approach is making sense. I still don't think the "internal tag" makes sense when we could just use the tag from the `enum`? I feel like it'll end up as the same amount of data because you still have the field `"user"` or `"project"` to de/serialise.

Reply:
I checked out the branch and I can see what you mean by the de/serialisation not being able to pick. I've tried https://serde.rs/enum-representations.html#adjacently-tagged which doesn't work and that baffles me...

In good news, I tried removing `EntityKind` and it looks like it's safe to remove.

Reply:
> I still don't think the "internal tag" makes sense when we could just use the tag from the `enum`? I feel like it'll end up as the same amount of data because you still have the field `"user"` or `"project"` to de/serialise.

The point is that the specific `User` and `Project` types are not an enum, and therefore do not generate any tag on their own.

Without those two properties the blobs generated serializing the specific types would not be deserializable by the enum type.

I understand this is tricky, but I had this requirement: being able to deserialize the **same** blob either as its specific type, or as the enum type.

To obtain this there are two hard sub-requirements:

* both the specific types and the enum must produce exactly the same blob
* the blob **must** contain "something" that allows the enum to distinguish the branches

That "something in the blob" are those two fields.

They are in the specific types so that they will be emitted anyway, whatever type is used to serialize.
And using the `#[serde(untagged)]` directive the enum serializes **exactly** as the internal specific types.

I could not find another serde setup that has the same properties... as I said, I tried, but it did not work.

-------------------

Now, for the sake of argument, there is another approach but I would find it even worse.

We could define `UserInfo` and `ProjectInfo` as "one branch enums", like

```
Enum UserInfo {
  User {
    // all UserInfo fields here...
  }
}
```

and the same for project.

Then I guess that their serialization format would be "naturally" compatible with `EntityInfo`, which contains both branches.

But i hope that you agree that this is horrible! :smiling_imp:
Those two zero-sized fields look like the lesser evil to me.

Reply:
Ya, I'm realising that this a PITA due to serde and it doesn't behave the way I expected. I was expecting behaviour such as the adjacent tags above.

Thanks for your patience with my qualms :) The approach is starting to make more sense to me now.

Reply:
@FintanH I guess what you're looking for is an "open sum" type:

```rust
enum Entity {
    User(User),
    Project(Project),
    Unknown(serde_json::Value),
}
```

I don't know if `serde` could be taught to handle this, nor how efficient it would be (does `serde_json` go through an intermediate representation if it can avoid it?). Also, `Unknown` should not be `Serialize`, really...

Reply:
Hmmm ya, that could be. It's weird though that it can't even serialise with this tag pragma. It complains that `type` is missing for `Entity`. I haven't actually written any serde code so I gave up fast :shrug: